lib/services/plugins-service.ts

```diff
@@ -0,0 +1,190 @@
+		return (() => {
+			let command = this.composeNpmCommand(npmCommandName, npmCommandArguments);
+			let result = this.$childProcess.exec(command, { cwd: this.$projectData.projectDir }).wait();
+			return this.parseNpmCommandResult(result);
+		}).future<string>()();
+	}
+	
+	private composeNpmCommand(npmCommandName: string, npmCommandArguments: string): string {
+		let command = `npm ${npmCommandName} ${npmCommandArguments} --save `;
+		if(this.$options.production) {
+			command += " --production ";
+		}
+		
+		return command;
+	}
+	
+	private parseNpmCommandResult(npmCommandResult: string): string { // The npmCommandResult is in the following format: [<name>@<version node_modules/<name>]
+		return npmCommandResult.split("@")[0]; // returns plugin name
+	}
+	
+	private executeForAllInstalledPlatforms(action: (pluginDestinationPath: string, platform: string, platformData: IPlatformData) => void): void {
+		let availablePlatforms = _.keys(this.$platformsData.availablePlatforms);
+		_.each(availablePlatforms, platform => {
+			let isPlatformInstalled = this.$fs.exists(path.join(this.$projectData.platformsDir, platform.toLowerCase())).wait();
+			if(isPlatformInstalled) {
+				let platformData = this.$platformsData.getPlatformData(platform.toLowerCase());
+				let pluginDestinationPath = path.join(platformData.appDestinationDirectoryPath, constants.APP_FOLDER_NAME, "tns_modules");
+				action(pluginDestinationPath, platform.toLowerCase(), platformData);
+			}
+		});
+	}
+	
+	private getNodeModuleData(moduleName: string): INodeModuleData {
+		let fullNodeModulePath = path.join(this.nodeModulesPath, moduleName);
+		let packageJsonFilePath = path.join(fullNodeModulePath, "package.json");
+		let data = require(packageJsonFilePath);
+		let result = {
+			name: data.name,
+			version: data.version,
+			isPlugin: data.nativescript,		
+			fullPath: fullNodeModulePath,
+		};
+		
+		return result;
+	}
+	
+	private convertToPluginData(nodeModuleData: INodeModuleData): IPluginData {
+		let pluginData: any = _.extend({}, nodeModuleData);
+		let data = <any>(nodeModuleData.isPlugin);
+		if(data) {
+			pluginData.platformsData = data.platforms;
+		}
+		
+		return pluginData;
+	}
+	
+	private get nodeModulesPath(): string {
+		return path.join(this.$projectData.projectDir, "node_modules");
+	}	
+	
+	private getInstalledFrameworkVersion(platform: string): IFuture<string> {
+		return (() => {
+			let platformData = this.$platformsData.getPlatformData(platform);
+			this.$projectDataService.initialize(this.$projectData.projectDir);
+			let frameworkData = this.$projectDataService.getValue(platformData.frameworkPackageName).wait();
+			return frameworkData.version;
+		}).future<string>()();
+	}
+	
+	private mergeXml(xml1: string, xml2: string): IFuture<string> {
+		let future = new Future<string>();
+		xmlmerge.merge(xml1, xml2, "", (mergedXml: string) => { // TODO: process errors
```

Comment:
The API does not provide errors (https://github.com/heysdk/xmlmerge-js/blob/master/xmlmerge.js#L243) so we cannot handle them.